### DISCUSSION TITLE
1st Class Extensible Pattern Matching

### DISCUSSION THREAD

**Discussion by sirinath** [dotnet/csharplang#8765] [Language Ideas]
At the time of writing we might not be able to envision all the patterns to match. So it is possible to consider a way that pattern matching is extensible.

At a basic level it would be like adding new `case` statements to a `switch`.

**Comment by colejohnson66** [comment]
Can you give an example of the problem you’re encountering? And what a solution would look like?

**Reply by sirinath** [reply]
It would at minimum like partial functions in Scala.

But I thin C# can do better.

I am not sure how the feature should be designed for best language ergonomic in C# though.

**Reply by HaloFour** [reply]
I think you're going to have to be a bit more specific in what kind of problem you're trying to solve and how something akin to Scala partial functions would solve that problem.  Examples of syntax before and after would also be helpful, the latter can be purely illustrative and doesn't have to represent a proposed end state.  However, just saying that C# needs to do yadda, yadda, yadda better than Scala isn't actionable.

**Reply by sirinath** [reply]
Something like:

```
var x = partial z => { 
  z switch {
  1 => ...
  2 => ...
  }
}

var y = x orElse  partial z => { 
  z switch {
  3 => ...
  4 => ...
  }
}

var a = y(p)
```

This will be effectively:

```
z switch {
  1 => ...
  2 => ...
  3 => ...
  4 => ...
}
```

This might not be the most elegant way to have this in C#. This is just to illustrate what I mean.

**Reply by HaloFour** [reply]
What does it do?  How is it used?  What problems does it solve?

**Reply by sirinath** [reply]
> What does it do? 

In short you can add more cases to a switch.

> How is it used? 

There times that certain case statements in a switch are identical across multiple switch statements. These can be written once and used multiple times with the additional cases added to each statement as needed.

> What problems does it solve?

You can use common case statements while writing them just one which helps maintainability.

It is like classes for pattern matching or switch statements. I.e., you wite a base patten with base functionality and then extend it of add more patter matching capabilities.

**Comment by HaloFour** [comment]
#8612 ?
